tests/test_adaptive_support.py

```diff
@@ -0,0 +1,133 @@
+from unittest import TestCase
+
+import numpy as np
+import pytest
+
+from aspire.denoising import adaptive_support
+from aspire.image import Image
+from aspire.source import ArrayImageSource
+from aspire.utils import circ, gaussian_2d, hankel
+
+logger = logging.getLogger(__name__)
+
+DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
+
+
+class AdaptiveSupportTest(TestCase):
+    def setUp(self):
+
+        self.resolution = resolution = 1025
+        self.sigma = sigma = 128
+        self.n_disc = n_disc = 10
+
+        # Create 2D Gaussian as initial array.
+        discs = np.tile(
+            gaussian_2d(resolution, sigma_x=sigma, sigma_y=sigma),
+            (n_disc, 1, 1),
+        )
+
+        # Add varying radius of solid disc.
+        #  The Fourier transform will yield Airy disc
+        #  which has more interesting content in F space.
+        for d in range(n_disc):
+            discs[d] = discs[d] + circ(resolution, radius=(d + 1) ** 2)
+
+        self.img_src = ArrayImageSource(Image(discs))
+
+        # Reference thesholds
+        self.references = {
+            1: 0.68,
+            2: 0.96,
+            3: 0.999,  # slightly off
+            self.resolution / (2 * self.sigma): 1,
+        }
+
+    def testAdaptiveSupportBadThreshold(self):
+        """
+        Method should raise meaningful error when passed unreasonable thresholds.
+        """
+
+        with pytest.raises(ValueError, match=r"Given energy_threshold.*"):
+            _ = adaptive_support(self.img_src, -0.5)
+
+        with pytest.raises(ValueError, match=r"Given energy_threshold.*"):
+            _ = adaptive_support(self.img_src, 9000)
+
+    def testAdaptiveSupportIncorrectInput(self):
+        """
+        Method should raise meaningful error when passed wrong format input.
+        """
+
+        with pytest.raises(
+            RuntimeError,
+            match="adaptive_support expects `Source` instance or subclass.",
+        ):
+            # Pass numpy array.
+            _ = adaptive_support(np.empty((10, 32, 32)))
+
+    def test_adaptivate_support_hankel(self):
+        """
+        Test support of Hankel function in Real and Fourier space is similar.
+
+        The Fourier transform of `hankel` should be similar to real space,
+        so we can test the support is similar.
+        """
+
+        # Generate stack of Hankel function images.
+        resolution = 64
+        imgs = np.tile(
+            hankel(resolution),
+            (self.n_disc, 1, 1),
+        )
+        # Centered Fourier Transform
+        f_imgs = np.fft.fftshift(np.sqrt(np.abs(np.fft.fft2(imgs))))
+
+        # Setup ImageSource like objects
+        img_src = ArrayImageSource(Image(imgs))
+        f_img_src = ArrayImageSource(Image(f_imgs))
+
+        thresholds = list(self.references.values())
+
+        for threshold in thresholds:
+            _, r = adaptive_support(img_src, threshold)
+            _, rf = adaptive_support(f_img_src, threshold)
```

Comment:
I'm confused. Why not get `r` and `rf` from a single call to `adaptive_support` on `img_src`?

Reply:
They are not close enough to equivalent after normalization correction (`*size`) to form a unit test.  

I'll try to elaborate below in a related discussion.

Reply:
Ok closing this then.